### PR TITLE
✨(frontend) hide teacher dashboard link

### PR DESCRIPTION
**Pull request by rlecellier**
## Purpose

The teacher dashboard access must be limited to user that are linked to a course or an organization.

### CODE REVIEW

File: src/frontend/js/widgets/UserLogin/index.tsx [hunk 1]
```diff
@@ -72,11 +74,51 @@ const UserLogin = ({ profileUrls = {} }: UserLoginProps) => {
...
+  const userMenuItems = useMemo(() => {
+    if (isReady) {
+      const unauthorizedUrlFilter = (urlItem: UserMenuItem) => {
+        if (urlItem.key === 'dashboard_teacher') {
+          if (isAbilitiesFetched) {
+            return (
+              joanieUser.abilities[JoanieAbilities.HAS_ORGANIZATION_ACCESS] ||
+              joanieUser.abilities[JoanieAbilities.HAS_COURSE_ACCESS]
+            );
+          }
```
Comment (rlecellier) [inline src/frontend/js/widgets/UserLogin/index.tsx:96]: I'm thinking that this should move into a `Abitility` object that would be use as: 
```ts
Abilities.can(JoanieUser, Abilities.ACCESS_TEACHER_DASHBOARD)
```

File: src/frontend/js/hooks/useUser.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  queryKey: ['user'],
+  apiInterface: () => useJoanieApi().user.me,
+  session: true,
+  messages,
+};
+/**
+ * Joanie Api hook to retrieve the authenticated user.
+ */
+export const useUser = ({ enabled = true }: Pick<QueryOptions<JoanieUser>, 'enabled'> = {}) => {
+  return useResource(props)(undefined, {}, { enabled });
```
Comment (jbpenrath) [inline src/frontend/js/hooks/useUser.tsx:30]: I wonder if this is not easiest to use `useSessionQuery` directly for this specific query.

**Review by jbpenrath — COMMENTED** [review]
From my point of view, we should fetch user info as soon as possible (like order, address and credit cards)

File: src/frontend/js/utils/AbilitiesHelper/index.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+export const abilityActions = { ...JoanieUserActions };
+
+const AbilitiesHelper = {
+  can: (entity: Maybe<Nullable<Entity>>, action: Actions) => {
+    if (!entity) {
+      return false;
+    }
+
+    if (isJoanieUserEntity(entity)) {
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/utils/AbilitiesHelper/index.ts]: if we need it we would be able to add a 'contextEntities' option like: 
```ts
// index.ts
interface ContextEntities {
  joanieUser?: JoanieUser,
  courseLight?: CourseLight,
}
const can = (
  entity: Maybe<Nullable<Entity>>,
  action: Actions,
  contextEntities: ContextEntities,
) => {
  if (!entity) {
    return false;
  }

  if (isJoanieUserEntity(entity)) {
    return joanieUserAbilities[action](entity, contextEntities);
  }
// joanieUserAbilities.ts
const abilities: JoanieUserAbilityList = {
  [JoanieUserActions.CAN_EXAMPLE]: (joanieUser: JoanieUser, contextEntities: ContextEntities) => {
    if (!contextEntities.courseLight) {
      return false
    }
    return (
      Boolean(joanieUser.abilities[JoanieUserAbilityActions.HAS_DUMMY_ACCESS]) ||
      Boolean(courseLight.abilities[CourseLightAbilityActions.HAS_DUMMY_ACCESS]),
    )
  }
};
```

But for now, it's better to to it as simple as possible.

File: src/frontend/js/hooks/useUser.tsx (hunk 2 above)
Comment (rlecellier) [reply src/frontend/js/hooks/useUser.tsx]: Now that i moved it into `JoanieSessionProvider` i like to have this `useResource` to plug it the same way addresses, orders and creditCards are plugged.
The change on enable is no longer needed.

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/OrderStateMessage/index.tsx
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/OrderStateMessage/index.tsx:1]: Not quit sure to understand why this change fix the error described in the commit message.

File: src/frontend/js/utils/AbilitiesHelper/index.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import { Maybe, Nullable } from 'types/utils';
```
Comment (jbpenrath) [inline src/frontend/js/utils/AbilitiesHelper/index.ts:1]: In the commit message :

```diff
-engin that handle it.
+engine that handles it.
```

File: src/frontend/js/utils/AbilitiesHelper/types.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { CourseLight } from 'types/Joanie';
+import { JoanieUserProfile } from 'types/User';
+
+export enum JoanieUserActions {
+  CAN_ACCESS_TEACHER_DASHBOARD = 'can_access_teacher_dahboard',
+}
+export const isJoanieUserEntity = (entity: Entity): entity is JoanieUserProfile =>
+  JSON.stringify(Object.keys(entity as JoanieUserProfile).sort()) ===
+  JSON.stringify(['abilities', 'full_name', 'id', 'is_staff', 'is_superuser', 'username']);
```
Comment (jbpenrath) [inline src/frontend/js/utils/AbilitiesHelper/types.ts:9]: We tend to ban the use of arrow function without explicit return statement mainly to ease function refactoring and readability.

File: CHANGELOG.md [hunk 6]
```diff
@@ -9,13 +9,15 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 ## [Unrealeased]
 
 ### Added
-
+
```
Comment (jbpenrath) [inline CHANGELOG.md:12]: Remove this extra space

File: src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx [hunk 7]
```diff
@@ -113,11 +116,13 @@ const JoanieSessionProvider = ({ children }: React.PropsWithChildren<any>) => {
     () => ({
       user,
       isLoadingUser,
+      joanieUser: joanieUser.item,
+      isLoadingJoanieUser: joanieUser.states.fetching,
```
Comment (jbpenrath) [inline src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx:120]: Why not just exposing one fetching indicator ?

File: src/frontend/js/contexts/SessionContext/BaseSessionProvider.tsx [hunk 8]
```diff
@@ -70,6 +70,8 @@ const BaseSessionProvider = ({ children }: PropsWithChildren<any>) => {
     () => ({
       user,
       isLoadingUser,
+      joanieUser: null,
+      isLoadingJoanieUser: false,
```
Comment (jbpenrath) [inline src/frontend/js/contexts/SessionContext/BaseSessionProvider.tsx:74]: Hmmm... I do not like the idea to add joanie related properties into this provided which is supposed to fully ignored joanie.

File: src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx [hunk 9]
```diff
@@ -113,11 +116,13 @@ const JoanieSessionProvider = ({ children }: React.PropsWithChildren<any>) => {
     () => ({
       user,
       isLoadingUser,
+      joanieUser: joanieUser.item,
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx]: I renamed the type and forgot these vars / properties.

File: src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx (hunk 7 above)
Comment (rlecellier) [reply src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx]: I follow what's exist for User:
```ts
user,
isLoadingUser,
```

File: src/frontend/js/contexts/SessionContext/BaseSessionProvider.tsx (hunk 8 above)
Comment (rlecellier) [reply src/frontend/js/contexts/SessionContext/BaseSessionProvider.tsx]: You ask to move this into Session context, how can i do it without adding these properties to `SessionContext` interface ?

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/OrderStateMessage/index.tsx
Comment (rlecellier) [reply src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/OrderStateMessage/index.tsx:1]: `orderStatusMessages` do not include validated state but the sentry check use `orderStatusMessages` instead of `OrderState`.
So "a error was throw when we had "validated" state."

File: src/frontend/js/hooks/useJoanieUserProfile.tsx [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    id: 'hooks.useUser.errorGet',
+    description: 'Error message shown to the user when user fetch request fails.',
+    defaultMessage: 'An error occurred while fetching logged user. Please retry later.',
+  },
+  errorNotFound: {
+    id: 'hooks.useUser.errorNotFound',
+    description: "Error message shown to the user when it isn't logged.",
+    defaultMessage: "You aren't log in.",
+  },
+});
```
Comment (jbpenrath) [inline src/frontend/js/hooks/useJoanieUserProfile.tsx:18]:
```suggestion
  errorGet: {
    id: 'hooks.useJoanieUserProfile.errorGet',
    description: 'Error message shown to the user when user fetch request fails.',
    defaultMessage: 'An error occurred while fetching user profile information. Please retry later.',
  },
  errorNotFound: {
    id: 'hooks.useJoanieUserProfile.errorNotFound',
    description: "Error message shown to the user when it isn't logged.",
    defaultMessage: "You aren't log in.",
  },
});
```

File: src/frontend/js/hooks/useJoanieUserProfile.tsx [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+const props: UseResourcesProps<JoanieUserProfile, ResourcesQuery, API['user']['me']> = {
+  queryKey: ['user', 'joanie'],
```
Comment (jbpenrath) [inline src/frontend/js/hooks/useJoanieUserProfile.tsx:21]: As you set session to true, a sessionQuery is used under the hood, so no need to add `user` into the queryKey. Furthermore, from my point of view, `profile` is more explicit than `joanie`
```suggestion
  queryKey: ['profile'],
```

File: src/frontend/js/hooks/useUser.tsx [hunk 12]
```diff
@@ -0,0 +1,29 @@
+import { defineMessages } from 'react-intl';
```
Comment (jbpenrath) [inline src/frontend/js/hooks/useUser.tsx]: Looks redundant with `useJoanieUserProfile`, you should remove this hook

File: src/frontend/js/utils/AbilitiesHelper/index.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+import joanieUserProfileAbilities from './joanieUserProfileAbilities';
+import { Entity, Actions, JoanieUserProfileActions, isJoanieUserProfileEntity } from './types';
+
+// further actions can be add here
+// like: { ...JoanieUserProfileActions, ...CourseActions };
+export const abilityActions = { ...JoanieUserProfileActions };
+
+/**
+ * Check if an action is available on an entity
+ * Joanie's backend give use an "abilities" entry for several models like user, course and organization
```
Comment (jbpenrath) [inline src/frontend/js/utils/AbilitiesHelper/index.ts:11]:
```
 * Joanie's backend give us an "abilities" entry for several models like user, course and organization
```

File: src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        [JoanieUserApiAbilityActions.HAS_ORGANIZATION_ACCESS]: false,
+        [JoanieUserApiAbilityActions.HAS_COURSE_ACCESS]: true,
+      },
+    },
+    {
+      label: 'User liked to an organization',
+      abilities: {
+        [JoanieUserApiAbilityActions.HAS_ORGANIZATION_ACCESS]: true,
+        [JoanieUserApiAbilityActions.HAS_COURSE_ACCESS]: false,
+      },
```
Comment (jbpenrath) [inline src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts:27]:
```suggestion
      label: 'User has access to both a course and an organization',
      abilities: {
        [JoanieUserApiAbilityActions.HAS_ORGANIZATION_ACCESS]: true,
        [JoanieUserApiAbilityActions.HAS_COURSE_ACCESS]: true,
      },
    },
    {
      label: 'User has access to a course',
      abilities: {
        [JoanieUserApiAbilityActions.HAS_ORGANIZATION_ACCESS]: false,
        [JoanieUserApiAbilityActions.HAS_COURSE_ACCESS]: true,
      },
    },
    {
      label: 'User has access to an organization',
      abilities: {
        [JoanieUserApiAbilityActions.HAS_ORGANIZATION_ACCESS]: true,
        [JoanieUserApiAbilityActions.HAS_COURSE_ACCESS]: false,
      },
```

File: src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    }).one();
+    expect(
+      AbilitiesHelper.can(joanieUser, JoanieUserProfileActions.CAN_ACCESS_TEACHER_DASHBOARD),
+    ).toBeTruthy();
+    expect(
+      AbilitiesHelper.cannot(joanieUser, JoanieUserProfileActions.CAN_ACCESS_TEACHER_DASHBOARD),
+    ).toBeFalsy();
+  });
+
+  it("shouldn't allow ACCESS_TEACHER_DASHBOARD when user isn't link to neither course or organization", () => {
```
Comment (jbpenrath) [inline src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts:41]:
```suggestion
  it("shouldn't allow ACCESS_TEACHER_DASHBOARD when user has access to neither course or organization", () => {
```

File: src/frontend/js/utils/AbilitiesHelper/types.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import { JoanieUserProfile } from 'types/User';
+
+export enum JoanieUserProfileActions {
+  CAN_ACCESS_TEACHER_DASHBOARD = 'can_access_teacher_dahboard',
+}
+export const isJoanieUserProfileEntity = (entity: Entity): entity is JoanieUserProfile =>
+  JSON.stringify(Object.keys(entity as JoanieUserProfile).sort()) ===
+  JSON.stringify(['abilities', 'full_name', 'id', 'is_staff', 'is_superuser', 'username']);
```
Comment (jbpenrath) [inline src/frontend/js/utils/AbilitiesHelper/types.ts]: I'm not quit sure `sort` method works well if keys contains special chars.
```suggestion
const JOANIE_PROFILE_KEYS = ['abilities', 'full_name', 'id', 'is_staff', 'is_superuser', 'username'];

export const isJoanieUserProfileEntity = (entity: Entity): entity is JoanieUserProfile => {
  const entityKeys = Object.keys(entity);
  
  if (entityKeys.length !== JOANIE_PROFILE_KEYS.length) return false;
  return entityKeys.every((key) => JOANIE_PROFILE_KEYS.includes(key));
};
```

File: src/frontend/js/hooks/useUser.tsx [hunk 17]
```diff
@@ -1,29 +0,0 @@
-import { defineMessages } from 'react-intl';
```
Comment (jbpenrath) [inline src/frontend/js/hooks/useUser.tsx]: Ok this has been removed into the wrong commit.

File: src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx [hunk 18]
```diff
@@ -110,11 +113,21 @@ const JoanieSessionProvider = ({ children }: React.PropsWithChildren<{}>) => {
     () => ({
       user,
       isPending,
+      joanieUserProfile: joanieUserProfile.item,
+      isPendingJoanieUserProfile: joanieUserProfile.states.fetching,
```
Comment (jbpenrath) [inline src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx:117]: There is a hook dedicated to this information, let's use it where it is needed WDYT ?

File: src/frontend/js/utils/test/createTestQueryClient.ts [hunk 19]
```diff
@@ -2,11 +2,13 @@ import { DehydratedState, hydrate } from '@tanstack/react-query';
 import { UserFactory } from 'utils/test/factories/richie';
```
Comment (jbpenrath) [inline src/frontend/js/utils/test/createTestQueryClient.ts:2]: In our commit message, explain why please.

File: src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx (hunk 18 above)
Comment (rlecellier) [reply src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx:117]: @jbpenrath should i also remove `isPending` as it's not used in our code ?

Comment (jbpenrath) [reply src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx]: Actually, we know we are fetching user information as long as user is `undefined`.
So indeed we could remove this indicator, but as this is a little out of the scope of this PR, I'll do nothing

File: src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx [hunk 20]
```diff
@@ -119,11 +122,20 @@ const JoanieSessionProvider = ({ children }: React.PropsWithChildren<{}>) => {
...
       destroy,
       login,
       register,
     }),
-    [user, isPending, destroy, login, register],
+    [
+      user,
+      isPending,
+      joanieUserProfile.item,
+      joanieUserProfile.states.fetching,
```
Comment (jbpenrath) [inline src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx:134]:
```suggestion
```

Comment (jbpenrath) [reply src/frontend/js/contexts/SessionContext/JoanieSessionProvider.tsx:134]: I'll totally remove profile information from this context. In this way there is only one way to retrieve user profile information, the hook `useJoanieUserProfile`

**Comment by rlecellier** [comment]
New way to do it:
add a hook `useJoanieUserAbilities`.
This hook will either call `useJoanieUserProfile` and return user abiltiies or return a object with all abilities to false.
This hook will be used instead of useJoanieUserProfile in the code.

This way, richie instance that do not use joanie will not break.

File: src/frontend/js/hooks/useJoanieUserProfile.tsx [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+  errorGet: {
+    id: 'hooks.useJoanieUserProfile.errorGet',
+    description: 'Error message shown to the user when user fetch request fails.',
+    defaultMessage:
+      'An error occurred while fetching user profile information. Please retry later.',
+  },
+  errorNotFound: {
+    id: 'hooks.useJoanieUserProfile.errorNotFound',
+    description: "Error message shown to the user when it isn't logged.",
+    defaultMessage: "You aren't log in.",
```
Comment (NathanVss) [inline src/frontend/js/hooks/useJoanieUserProfile.tsx:17]: logged* in

File: src/frontend/js/hooks/useUser.tsx [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+const props: UseResourcesProps<JoanieUserProfile, ResourcesQuery, API['user']['me']> = {
+  queryKey: ['user', 'joanie'],
+  apiInterface: () => useJoanieApi().user.me,
+  session: true,
+  messages,
+};
+/**
+ * Joanie Api hook to retrieve the authenticated user.
+ */
+export const useUser = useResource(props);
```
Comment (NathanVss) [inline src/frontend/js/hooks/useUser.tsx]: It looks like it is duplicated from `useJoanieUserProfile`, I might be missing something.

File: src/frontend/js/utils/AbilitiesHelper/index.ts [hunk 23]
```diff
@@ -0,0 +1,46 @@
+import { Maybe, Nullable } from 'types/utils';
+import joanieUserProfileAbilities from './joanieUserProfileAbilities';
+import { Entity, Actions, JoanieUserProfileActions, isJoanieUserProfileEntity } from './types';
+
```
Comment (NathanVss) [inline src/frontend/js/utils/AbilitiesHelper/index.ts:4]: commit body: "an engine"*

File: src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+    }).one();
+    expect(
+      AbilitiesHelper.can(joanieUser, JoanieUserProfileActions.CAN_ACCESS_TEACHER_DASHBOARD),
+    ).toBeTruthy();
+    expect(
+      AbilitiesHelper.cannot(joanieUser, JoanieUserProfileActions.CAN_ACCESS_TEACHER_DASHBOARD),
+    ).toBeFalsy();
+  });
+
+  it("shouldn't allow ACCESS_TEACHER_DASHBOARD when user isn't has access to neither course or organization", () => {
```
Comment (NathanVss) [inline src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts:41]: Shouldn't allow access to the teacher dashboard when the user doesn't have access to either the course or the organization.*

File: src/frontend/js/widgets/UserLogin/index.tsx [hunk 25]
```diff
@@ -73,10 +75,47 @@ const UserLogin = ({ profileUrls = {} }: UserLoginProps) => {
...
   const { user, destroy, login, register } = useSession();
   const intl = useIntl();
+  const {
+    item: joanieUserProfile,
+    states: { isFetched: isJoanieUserProfileFetched },
+  } = useJoanieUserProfile();
+
+  const isAnonymous = user === null;
+  const isUserFetched = user !== undefined;
+  const isReady = isAnonymous || (isUserFetched && isJoanieUserProfileFetched);
```
Comment (NathanVss) [inline src/frontend/js/widgets/UserLogin/index.tsx:85]: Do this could work instead? 

```suggestion
  const isReady = isAnonymous || !!joanieUserProfile;
```

**Review by NathanVss — COMMENTED** [review]
I get an infinite loader, but I think I lack a bit of context on this review 🤔

Capture d’écran 2023-11-28 à 14 44 23

File: src/frontend/js/widgets/UserLogin/index.tsx [hunk 26]
```diff
@@ -75,23 +75,13 @@ const UserLogin = ({ profileUrls = {} }: UserLoginProps) => {
...
   const userMenuItems = useMemo(() => {
-    if (isReady) {
+    if (joanieUserAbilities !== null) {
       const unauthorizedUrlFilter = (urlItem: UserMenuItem) => {
         if (urlItem.key === 'dashboard_teacher') {
-          return AbilitiesHelper.can(
-            joanieUserProfile,
-            abilityActions.CAN_ACCESS_TEACHER_DASHBOARD,
-          );
+          return joanieUserAbilities!.can(abilityActions.ACCESS_TEACHER_DASHBOARD);
```
Comment (jbpenrath) [inline src/frontend/js/widgets/UserLogin/index.tsx:84]: With the check above `!` is useless I think.

```suggestion
          return joanieUserAbilities.can(abilityActions.ACCESS_TEACHER_DASHBOARD);
```

File: src/frontend/js/widgets/UserLogin/index.spec.tsx [hunk 27]
```diff
@@ -103,13 +108,59 @@ describe('<UserLogin />', () => {
...
+
+    screen.getByText(user.username);
+    const settingsLink = screen.getByRole('link', { name: 'Settings' });
+    const accountLink = screen.getByRole('link', { name: 'Account' });
+    expect(settingsLink.getAttribute('href')).toEqual('https://auth.local.test/settings');
+    expect(accountLink.getAttribute('href')).toEqual(`https://auth.local.test/u/${user.username}`);
+    expect(screen.queryByRole('link', { name: 'Teacher dashboard' })).not.toBeInTheDocument();
+  });
+
+  it('should renders teacher dashboard menu item for user attached to an organization', async () => {
```
Comment (jbpenrath) [inline src/frontend/js/widgets/UserLogin/index.spec.tsx:148]:
```
  it('should renders teacher dashboard menu item for user with organization access', async () => {
```

File: CHANGELOG.md [hunk 28]
```diff
@@ -23,6 +23,8 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Changed
 
+- Teacher dashboard access is restricted to user that are linked to a course 
+  or an organization.
```
Comment (jbpenrath) [inline CHANGELOG.md:27]:
```suggestion
- Teacher dashboard access is restricted to user which has access to a course 
  or an organization.
```

File: src/frontend/js/hooks/useJoanieUserProfile.tsx [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+const props: UseResourcesProps<JoanieUserProfile, ResourcesQuery, API['user']['me']> = {
+  queryKey: ['profile'],
+  apiInterface: () => useJoanieApi().user.me,
+  session: true,
+  messages,
+};
+/**
+ * Joanie Api hook to retrieve the authenticated user.
+ */
+export const useJoanieUserProfile = () => useResource(props)(undefined, {}, { enabled: true });
```
Comment (jbpenrath) [inline src/frontend/js/hooks/useJoanieUserProfile.tsx:30]: We could enable this query only when user is authenticated.

```suggestion
export const useJoanieUserProfile = () => useResource(props)(undefined, {}, { enabled: undefined });
```

Or in a way more explicit : 

```suggestion
export const useJoanieUserProfile = () => {
	const { user } = useSession()
	return useResource(props)(undefined, {}, { enabled: !!user });
}
```

File: src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts (hunk 24 above)
Comment (jbpenrath) [reply src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts:41]: The typo is still here

```suggestion
  it("shouldn't allow access to the teacher dashboard when the user doesn't have access to either the course or the organization", () => {
```

File: src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+  ])('should allow ACCESS_TEACHER_DASHBOARD when "$label"', ({ abilities }) => {
+    const joanieUser = JoanieUserProfileFactory({
+      abilities,
+    }).one();
+    expect(
+      AbilitiesHelper.can(joanieUser, JoanieUserProfileActions.ACCESS_TEACHER_DASHBOARD),
+    ).toBeTruthy();
+    expect(
+      AbilitiesHelper.cannot(joanieUser, JoanieUserProfileActions.ACCESS_TEACHER_DASHBOARD),
+    ).toBeFalsy();
```
Comment (jbpenrath) [inline src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts:38]: `toBeTruthy` or `toBeFalsy` can lead to false positive.
 We tried to remove the use of those methods in our test in favor of `toBe(true)` and `toBe(false)`

File: src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts [hunk 31]
```diff
@@ -0,0 +1,55 @@
...
+        [JoanieUserApiAbilityActions.HAS_ORGANIZATION_ACCESS]: false,
+        [JoanieUserApiAbilityActions.HAS_COURSE_ACCESS]: false,
+      },
+    }).one();
+    expect(
+      AbilitiesHelper.can(joanieUserProfile, JoanieUserProfileActions.ACCESS_TEACHER_DASHBOARD),
+    ).toBeFalsy();
+    expect(
+      AbilitiesHelper.cannot(joanieUserProfile, JoanieUserProfileActions.ACCESS_TEACHER_DASHBOARD),
+    ).toBeTruthy();
```
Comment (jbpenrath) [inline src/frontend/js/utils/AbilitiesHelper/joanieUserProfileAbilities.spec.ts:53]: Ditto use `toBe(true)` and `toBe(false)` instead.

File: src/frontend/js/widgets/UserLogin/index.spec.tsx [hunk 32]
```diff
@@ -30,30 +33,50 @@ jest.mock('utils/indirection/window', () => ({
...
+    </QueryClientProvider>
+  );
+  beforeEach(() => {
+    fetchMock
+      .get('https://endpoint.test/api/v1.0/orders/', [])
+      .get('https://endpoint.test/api/v1.0/addresses/', [])
+      .get('https://endpoint.test/api/v1.0/credit-cards/', []);
+  });
   afterEach(() => {
+    jest.clearAllMocks();
```
Comment (jbpenrath) [inline src/frontend/js/widgets/UserLogin/index.spec.tsx:65]: We do not mock new things, I do not understand why we have to add this `jest.clearAllMocks()`

File: src/frontend/js/hooks/useJoanieUserAbilities/index.tsx [hunk 33]
```diff
@@ -0,0 +1,23 @@
...
+export const useJoanieUserAbilities = () => {
+  if (!isJoanieEnabled) {
+    return AbilitiesHelper.buildEntityInterface({
+      id: 'ANONYMOUS',
+      username: 'anonymous_user',
+      full_name: 'Anonymous user',
+      is_superuser: false,
+      is_staff: false,
+      abilities: {},
+    });
```
Comment (jbpenrath) [inline src/frontend/js/hooks/useJoanieUserAbilities/index.tsx:14]: We could just return `undefined` WDYT?

```
undefined === "Not configured"
null === "Fetching..."
Object === "Fetched"
```

File: src/frontend/js/hooks/useJoanieUserProfile.tsx (hunk 29 above)
Comment (rlecellier) [reply src/frontend/js/hooks/useJoanieUserProfile.tsx:30]: i first did that and it required to code useJoanieUserAbilities differently.
I choose this way to do it but i'll change if your don't like it.

File: src/frontend/js/hooks/useJoanieUserAbilities/index.tsx (hunk 33 above)
Comment (rlecellier) [reply src/frontend/js/hooks/useJoanieUserAbilities/index.tsx:14]: I prefer to have a anonymous object.
but i'll change it if you don't like it this way.

File: src/frontend/js/widgets/UserLogin/index.spec.tsx [hunk 34]
```diff
@@ -105,6 +104,7 @@ describe('<UserLogin />', () => {
     expect(await screen.findByRole('button', { name: 'Log in' })).toBeInTheDocument();
     expect(screen.getByRole('button', { name: 'Sign up' })).toBeInTheDocument();
 
+    // await waitForElementToBeRemoved(screen.queryByText('Loading login status...'));
```
Comment (jbpenrath) [inline src/frontend/js/widgets/UserLogin/index.spec.tsx:107]: Debug residue ?

File: src/frontend/js/widgets/UserLogin/index.tsx [hunk 35]
```diff
@@ -73,10 +75,37 @@ const UserLogin = ({ profileUrls = {} }: UserLoginProps) => {
    */
   const { user, destroy, login, register } = useSession();
   const intl = useIntl();
+  const joanieUserAbilities = useJoanieUserAbilities();
+
+  const userMenuItems = useMemo(() => {
+    if (joanieUserAbilities !== null) {
```
Comment (jbpenrath) [inline src/frontend/js/widgets/UserLogin/index.tsx:81]: I feel now joanieUserAbilities will never return null

File: src/frontend/js/widgets/UserLogin/index.tsx [hunk 36]
```diff
@@ -73,10 +75,40 @@ const UserLogin = ({ profileUrls = {} }: UserLoginProps) => {
...
+        },
+      ];
+    }
+    return [];
+  }, [user, joanieUserAbilities]);
 
   return (
     <div className="user-login">
-      {user === undefined ? (
+      {user === undefined || joanieUserAbilities === null ? (
```
Comment (jbpenrath) [inline src/frontend/js/widgets/UserLogin/index.tsx:111]: I still do not understand when joanieUserAbilities can be equal to null with the current logic.

**Review by jbpenrath — APPROVED** [review]
LGTM ✅
